src/base.rs

```diff
@@ -466,63 +422,200 @@ impl fmt::Display for Location {
+        inner: FactStructure,
+    },
     /// Splitter
     Splitter,
 }
 
 #[allow(missing_docs)]
 impl Fact {
     pub fn new<K: Into<String>, V: Into<String>>(key: K, value: V) -> Fact {
-        Fact::KeyValue {
-            key: key.into(),
-            value: value.into(),
+        Fact::Structural {
+            inner: FactStructure::KeyValue {
+                key: key.into(),
+                value: Box::new(FactStructure::Value {
+                    formatted_value: value.into(),
+                }),
+            },
         }
     }
     pub fn new_simple_fact<V: Into<String>>(value: V) -> Fact {
-        Fact::Value {
-            value: value.into(),
+        Fact::Structural {
+            inner: FactStructure::Value {
+                formatted_value: value.into(),
+            },
         }
     }
     pub fn new_multi_value_fact<K: Into<String>, V: Into<String>>(key: K, values: Vec<V>) -> Fact {
-        Fact::KeyValues {
-            key: key.into(),
-            values: values.into_iter().map(|v| v.into()).collect(),
+        Fact::Structural {
+            inner: FactStructure::KeyValue {
+                key: key.into(),
+                value: Box::new(FactStructure::List {
+                    values: values
+                        .into_iter()
+                        .map(|v| {
+                            Box::new(FactStructure::Value {
+                                formatted_value: v.into(),
+                            })
+                        })
+                        .collect(),
+                }),
+            },
         }
     }
     pub fn new_splitter() -> Fact {
         Fact::Splitter
     }
 }
 
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum FactStructure {
+    /// Single assertion message
+    ///
+    /// # Example
+    /// ```text
+    /// FactStructure::Value { formatted_value: "expected that the vec is empty" }
+    /// ```
+    Value { formatted_value: String },
+    /// Assertion message for multiple values
+    ///
+    /// # Example
+    /// ```text
+    /// FactStructure::List {
+    ///   values: vec![
+    ///     FactStructure::Value { formatted_value: "foo" },
+    ///     FactStructure::Value { formatted_value: "bar" }
+    ///   ]
+    /// }
+    /// ```
+    List { values: Vec<Box<FactStructure>> },
+    /// Keyed assertion message
+    ///
+    /// # Example
+    /// ```text
+    /// FactStructure::KeyValue {key: "expected", value: FactStructure::Value { formatted_value: "foo" }}
+    /// ```
+    KeyValue {
+        key: String,
+        value: Box<FactStructure>,
+    },
+    /// Keyed assertion structured message
+    ///
+    /// # Example
+    /// ```text
+    /// FactStructure::Nested {
+    ///   inner: vec![
+    ///     ("expected", FactStructure::Value { formatted_value: "foo" }),
+    ///     ("actual", FactStructure::Value { formatted_value: "bar" })
+    ///   ]
+    /// }
+    /// ```
+    Nested {
+        inner: Vec<(String, Box<FactStructure>)>,
```

Comment:
Is there any possiblility of creating two or more nested facts?  For example, 
```
"first": {"second": {"actual": "xxx", "expected": "yyy" }}
```

This should not happen to avoid too nested output, and I guess this doesn't happen.  If we want to do this like in proto or json, we should use path rather than structured output  to repdesent where making dif like `modified: parent.child.name: actual -> expected`.  This means generalization of nest is probably overengineering, as one nested should be sufficient.

It is one option to add a enum entry that doesn't need nest for example `Fact::KeyedDiff {key: String, actual: String, expected: String}`, which is close to MapFactDiff.  I'm wondering if this is sufficient to fix the original bug or not. 

Additionally, I'm planing to introduce the color diff feature in the future on #29.  In this feature, it is high likely to modify the fact structure to hold diff instead of KeyValue for example `Fact::Diff {actual: String, expected: String}`, which enables to colorize the message without changing the generation logic of facts.  KeyedDiff impl idea fits this approach, I think.  Do you have any thoughts on this? Thanks!